### ISSUE TITLE
SDL_Window fullscreen flag gets out of sync when fullscreen is set using the button from the window title bar on macOS

### ISSUE THREAD

**Issue by dortamiguel** [libsdl-org/SDL#8515]
I noticed that when a window is set to fullscreen mode on macOS using the button in the title bar, SDL won't set the internal flags of the window as fullscreen
![image](https://github.com/libsdl-org/SDL/assets/3449006/fe4e2754-342b-4262-a91f-c6727cd9e660)

The problem with this is that after going fullscreen using that way then a call to `SDL_SetWindowFullscreen(window, SDL_FALSE);` next time won't work since SDL will think that the window is not in fullscreen.

Now to fix this what I'm doing is checking the flags of the native window, and if I notice that SDL is out of sync then I will manually call SDL_SetWindowFullscreen with the right params. 

```c++
#include <Cocoa/Cocoa.h>

bool is_fullscreen_mode(SDL_Window *sdl_window) {
    NSWindow *window = (NSWindow *)CFBridgingRelease(
        SDL_GetProperty(SDL_GetWindowProperties(sdl_window), "SDL.window.cocoa.window")
    );
    return (window.styleMask & NSFullScreenWindowMask) == NSFullScreenWindowMask;
}

// On the event loop
#if __APPLE__
    if (event.type == SDL_EVENT_WINDOW_RESIZED) {
        if (is_fullscreen_mode(window)) {
            if (screen_mode != Screen_Mode::FULLSCREEN) {
                SDL_SetWindowFullscreen(window, SDL_TRUE);
            }
            screen_mode = Screen_Mode::FULLSCREEN;
        } else {
            if (screen_mode == Screen_Mode::FULLSCREEN) {
                SDL_SetWindowFullscreen(window, SDL_FALSE);
            }
            screen_mode = Screen_Mode::WINDOWED;
        }
    }
#endif
```

This solves the issue for me but, should this be considered a bug or maybe I'm doing somethign wrong?

**Comment by slime73** [comment]
Right now SDL effectively treats that button as "maximized in its own space", which has a bit different semantics from app-requested fullscreen that games typically use. The former doesn't promise that system UI will be hidden whereas the latter does, for example.

**Comment by dortamiguel** [comment]
should then SDL avoid allowing this button at all?

```
NSWindow *window = (NSWindow *)CFBridgingRelease(
    SDL_GetProperty(SDL_GetWindowProperties(sdl_window), "SDL.window.cocoa.window")
);
window.collectionBehavior = NSWindowCollectionBehaviorFullScreenNone;
```

**Comment by slime73** [comment]
The button is available when the window is created with the resizable flag, I believe. It's treated the same as other forms of user maximization of the window (and the other main type is available when the user holds `alt` and presses the button).

**Comment by dortamiguel** [comment]
To me it looks confusing that the fullscreen button doesn't toggle fullscreen inside SDL, but I guess that if this an intentional behaviour in SDL then this is not a bug?

**Comment by Kontrabant** [comment]
This fullscreen state flag discrepancy is also present in the X11 backend, which allows changing the fullscreen state via a system hotkey, as well. If we eventually end up taking #8472 in, it maintains the proper flag state when both entering/leaving fullscreen programmatically as well as via a system initiated event, since it changes it in response to system events vs proactively setting it as is currently done in the main branch.

**Comment by slime73** [comment]
I don't think this is related at all to #8472, is it? This is more of a design choice because the OS 'fullscreen' button on macOS is more like a maximize button (doesn't remove some system UI, etc) so SDL intentionally doesn't treat it as real fullscreen.

**Comment by slime73** [comment]
Does that mean an app needs to manually duplicate fullscreen query code now and compare the difference with what SDL's getter APIs return, in order to tell whether the app is in real desktop-fullscreen without UI obstructions the user will run into when the mouse over parts of the screen?
